### PR TITLE
More Precise Generated Types

### PR DESCRIPTION
**Pull request by brendanhay**
Since Go is better than Haskell I have added a little secret sauce to extract Terraform's type information into a more palatable form.

The result is now all nested settings types and additional information such as required vs optional, conflicts with, min/max items, and others are elaborated into Haskell datatypes and values with more precision, obviating the need to continually add additoinal type declarations by hand.

Unfortunately Terraform is still highly string oriented, so there is a need to add annotations to the  configuration per provider to guide the generator to emit more precise field types like `AWS.Region`, etc. But overall it seems like quite a large improvement.

Some points of interest:

* [./terrafomo-go/main.go](https://github.com/brendanhay/terrafomo/blob/568d3a583e1cad1dbb1c242f0f9e86afb9e19ecc/terrafomo-go/main.go) Is the very brief Go extraction code.
* [./terrafomo-gen/model](https://github.com/brendanhay/terrafomo/blob/568d3a583e1cad1dbb1c242f0f9e86afb9e19ecc/terrafomo-gen/model) Contains the AST extracted from the `terraform-provider-*` Go code.
* [./terrafomo-gen/ir](https://github.com/brendanhay/terrafomo/blob/568d3a583e1cad1dbb1c242f0f9e86afb9e19ecc/terrafomo-gen/ir) Contains the intermediate representation, the elaborated Haskell AST pre-rendering.
* [./terrafomo-gen/config](https://github.com/brendanhay/terrafomo/blob/568d3a583e1cad1dbb1c242f0f9e86afb9e19ecc/terrafomo-gen/config) As usual contains the per-provider (as Terraform names them) configuration. The existence of these configuration files is used by the `Makefile` to invoke the generator.

The comprehensive configuration format is demonstrated below, but most fields are optional:

```yaml
# Required: the name of the resulting Cabal package.
package-name: terrafomo-aws

# Required: the Haskell-cased provider name, used for module namespaces.
provider-name: AWS

# Optional: The maximum graph partition size - rough number of types
# per Resource/DataSource0N.hs module.
partition-size: 80

# Optional: Extra dependencies to list in the package.yaml
extra-deps:
  - amazonka-core
  - formatting
  - iproute

# Optional: Match a suffix, prefix, or exact field name and replace the type
# with the one specified. Affects _all_ fields, in the global sense, and uses
# The _original_ name from the terraform code.
field-overrides:
  - suffix: _cidr_block
    type: TF.Attr s P.Foo

  - prefix: zones_
    type: TF.Attr s P.Zone

  - exact: region
    type: TF.Attr s P.Region

# Optional: Match a data type name with exact field names to override.
# Occurs before field-overrides and short-circuits.
datatype-overrides:
  provider:
    region: P.Region
```

A contentious point. still up for discussion:

The generator now utilises the `conflictsWith` information to generate lens setters that ensure the emitted HCL code is always valid. This means if you set one field, and subsequently set a conflicting field - the first field will be 'reset'.

### CODE REVIEW

**Comment by brendanhay** [comment]
@nhibberd @thumphries

**Comment by thumphries** [comment]
First of all this is awesome and will save us a fair bit of toil, thanks!

> The generator now utilises the conflictsWith information to generate lens setters that ensure the emitted HCL code is always valid. This means if you set one field, and subsequently set a conflicting field - the first field will be 'reset'.

IIRC there is already a Reader / Writer thing going on in the Terraform monad, is there any possibility of it also throwing an error at render time when the user adds conflicts? Silent failure here scares me a bit.

(Would be happy to do the work for that)

**Comment by brendanhay** [comment]
The implementation of the validators (via the `Terraform` monad), is outlined in the following:

* [aws_s3_bucket](https://github.com/brendanhay/terrafomo/blob/41c2a81fabc6194a6833a487760a553edc3f54d4/provider/terrafomo-aws/gen/Terrafomo/AWS/Resource04.hs#L7665-L7690)
* [Terraform/Validator.hs](https://github.com/brendanhay/terrafomo/blob/41c2a81fabc6194a6833a487760a553edc3f54d4/terrafomo/src/Terrafomo/Validator.hs)

Basically every setting or resource schema gets an `IsValid` instance which providers a validator that can be combined by parent (containing) types. It is checked when declaring a resource in the `Terraform` monad and results in a (currently unstructured) error message containing all of the conflicting fields, morally similar to `Data.Either.Validation`.

**Comment by nhibberd** [comment]
You've dropped `provider/terrafomo-builtins/package.yaml`
